### DISCUSSION TITLE
Requesting feedback: GraphQL Subscription JS package design for use with ports

### DISCUSSION THREAD

**Discussion by dillonkearns** [dillonkearns/elm-graphql#558] [Ideas]
I would love feedback from anyone who is using GraphQL Subscriptions, or interested in using Subscriptions, who could give me some feedback based on their use case.

My goal is to publish a very simple JavaScript/TypeScript library that can be a community standard (that community members can contribute to and improve) that makes it easy to connect up GraphQL Subscriptions through a couple of ports.

Here are some rough notes for my current thinking:

* Users create their own simple `port module GraphqlSubscription`, and only expose a function which takes a GraphQL Subscription `SelectionSet`, and gives you back a `Cmd` and `Sub` that you can register to wire up GraphQL Subscriptions (see https://github.com/IncrementalElm/postgraphile-pomodoro/blob/master/client/src/GraphqlSubscription.elm)
* Note that the module doesn't expose the ports. Just a detail, but I think that's a good practice.
* You pass in the ports for starting the subscription, and sending subscription payloads back to elm, like so: https://github.com/IncrementalElm/postgraphile-pomodoro/blob/d7b43a1eaa9c1576feef26c921b26ef3653fb677/client/index.ts#L6-L10

My thoughts on the JavaScript/TypeScript module:
* It would be an NPM package that you can install
* Should there be an API for unsubscribing? What is the use case?
* Ideally the API should be as declarative as possible (the JavaScript code should transparently handle the details of reopening the connection and handling errors, and only tell the elm code about issues which it cannot handle. Otherwise the low-level details should be entirely abstracted away).
* Is it possible to handle multiple GraphQL frameworks for the handshake/protocol details for subscriptions? It would be nice to test this out and see if we can make it a general-purpose Subscription client.
* Perhaps the design should associate the GraphQL Subscription query document with how payloads are sent back... so when you get the `cmdAndSub`, you will only get incoming GraphQL Subscription Payloads for the exact query that you sent out when you called that.

Any and all feedback is appreciated!

**Comment by karthikvt26** [comment]
Looks great :) Here is what I felt based on my experience working with apollo-style-subscriptions 

1) Correct me if I am wrong. If I would like to subscribe to the results of two queries (`Query A` and `Query B`). Will I have to implement them seperately and construct two seperate `GraphqlSubscriptions` on the javascript side? If not is there a way using which I will be able to distinguish between the payloads (unique id to the query etc)

2) There should be an API to allow users to configure the headers (Authorization token etc). 

3) If I am building an order processing application, I would like to unsubscribe from the subscription once the backend reaches a consistent state. 

For example: Lets say I am placing an order. The state the order can be are as follows

1) Order Received
2) Order Confirmed
3) Order Dispatched
4) Order Delivered

If the front end is subscribing to this table. It should be able to unsubscribe once all of above turns `true`

Hope this is useful :) kindly let me know what you think?

**Comment by ropwareJB** [comment]
Currently using elm-graphql with backend in Absinthe.

I definitely think there is a good use case for unsubscribing - lets say I have a list of `User`, and users have `Project`s and can share them with one another. Upon viewing a page dedicated to a `Project`, you want to subscribe to modifications on that project, and any comments on that project by any user. If you navigate to another `Project` page altogether, you probably want to unsub to the previous and sub to the new one (Think of the situation of ~10min in, when a user has been viewing various object types and instances, the subscription list would pile up).

I'd imagine that given a subscription, serialize its arguments and that would give you a unique identifier 
 for that subscription stream. This could avoid having to supply subscription 'names' for subbing/unsubbing.
